File: src/main/java/toast/client/modules/movement/Velocity.java
package toast.client.modules.movement;

import net.minecraft.network.packet.s2c.play.EntityVelocityUpdateS2CPacket;
import net.minecraft.network.packet.s2c.play.ExplosionS2CPacket;
import toast.client.event.EventImpl;
import toast.client.event.events.network.EventPacketReceived;
import toast.client.modules.Module;

public class Velocity extends Module {
    public double lastx = 0;
    public double lasty = 0;
    public double lastz = 0;

    public Velocity() {
        super("Velocity", "Changes your velocity, can stop things from pushing you.", Module.Category.MOVEMENT, -1);
        this.settings.addMode("Mode", "Vanilla", "Vanilla", "YTeleport");
        this.settings.addBoolean("Cancel", true);
        this.settings.addSlider("HVel", -100.0, 0.0, 100.0);
        this.settings.addSlider("VVel", -100.0, 0.0, 100.0);
    }

    @Override
    public void onEnable() {
        this.lastx = 0D;
        this.lasty = 0D;
        this.lastz = 0D;
    }

    @Override
    public void onDisable() {
        this.lastx = 0D;
        this.lasty = 0D;
        this.lastz = 0D;
    }


    @EventImpl
    public void onPacket(EventPacketReceived e) {
        if (this.isEnabled()) {
            if (mc.player == null) return;
            if (this.isMode("Vanilla")) {
                if (e.getPacket() instanceof ExplosionS2CPacket) {
                    ExplosionS2CPacket s27 = (ExplosionS2CPacket) e.getPacket();
                    if (!this.getBool("Cancel")) {
                        double hvel = this.getDouble("HVel");
                        double vvel = this.getDouble("VVel");
                        mc.player.setVelocity(s27.getX() * (hvel * 0.00001),
                                s27.getY() * (vvel * 0.00001), s27.getZ() * (hvel * 0.00001));
                    }
                    e.setCancelled(true);
                }
                if (e.getPacket() instanceof EntityVelocityUpdateS2CPacket) {
                    EntityVelocityUpdateS2CPacket s12 = (EntityVelocityUpdateS2CPacket) e.getPacket();
                    if (s12.getId() != mc.player.getEntityId()) return;

                    if (!this.getBool("Cancel")) {
                        double hvel = this.getDouble("HVel");
                        double vvel = this.getDouble("VVel");
                        mc.player.setVelocity(s12.getVelocityX() * (hvel * 0.00001),
                                s12.getVelocityY() * (vvel * 0.00001), s12.getVelocityZ() * (hvel * 0.00001));
                    }
                    e.setCancelled(true);
                }
            } else if (this.isMode("YTeleport")) {
                if (mc.player.hurtTime == 0F) {
                    this.lastx = mc.player.getX();
                    this.lasty = mc.player.getY();
                    this.lastz = mc.player.getZ();
                } else if (mc.player.hurtTime >= 8) {
                    mc.player.setPos(this.lastx, this.lasty + 50D, this.lastz);
                }
            }
        }
    }
}

File: src/main/java/toast/client/modules/movement/Velocity.kt
package toast.client.modules.movement

import com.google.common.eventbus.Subscribe
import net.minecraft.network.packet.s2c.play.EntityVelocityUpdateS2CPacket
import net.minecraft.network.packet.s2c.play.ExplosionS2CPacket
import toast.client.events.network.EventPacketReceived
import toast.client.modules.Module

class Velocity : Module("Velocity", "Changes your velocity, can stop things from pushing you.", Category.MOVEMENT, -1) {
    var lastx = 0.0
    var lasty = 0.0
    var lastz = 0.0
    override fun onEnable() {
        lastx = 0.0
        lasty = 0.0
        lastz = 0.0
    }

    override fun onDisable() {
        lastx = 0.0
        lasty = 0.0
        lastz = 0.0
    }

    @Subscribe
    fun onPacket(e: EventPacketReceived) {
        if (this.enabled) {
            if (mc.player == null) return
            when (mode) {
                "Vanilla" -> when (e.getPacket()) {
                    is ExplosionS2CPacket -> {
                        val s27 = e.getPacket() as ExplosionS2CPacket
                        if (!getBool("Cancel")) {
                            val hvel = getDouble("HVel")
                            val vvel = getDouble("VVel")
                            mc.player!!.setVelocity(s27.x * (hvel * 0.00001),
                                    s27.y * (vvel * 0.00001), s27.z * (hvel * 0.00001))
                        }
                        e.isCancelled = true
                    }
                    is EntityVelocityUpdateS2CPacket -> {
                        val packetEntityId = e.getPacket() as EntityVelocityUpdateS2CPacket
                        if (packetEntityId.id != mc.player!!.entityId) return
                        if (!getBool("Cancel")) {
                            val hVel = getDouble("HVel")
                            val vVel = getDouble("VVel")
                            mc.player!!.setVelocity(packetEntityId.velocityX * (hVel * 0.00001),
                                    packetEntityId.velocityY * (vVel * 0.00001), packetEntityId.velocityZ * (hVel * 0.00001))
                        }
                        e.isCancelled = true
                    }
                }
                "YTeleport" -> when {
                    mc.player!!.hurtTime.toFloat() == 0f -> {
                        lastx = mc.player!!.x
                        lasty = mc.player!!.y
                        lastz = mc.player!!.z
                    }
                    mc.player!!.hurtTime >= 8 -> mc.player!!.setPos(lastx, lasty + 50.0, lastz)
                }
            }
        }
    }

    init {
        settings.addMode("Mode", "Vanilla", "Vanilla", "YTeleport")
        settings.addBoolean("Cancel", true)
        settings.addSlider("HVel", -100.0, 0.0, 100.0)
        settings.addSlider("VVel", -100.0, 0.0, 100.0)
    }
}